File: Day15/solution.js
// Changed seen from Map() to Uint32Array(), which is apparently way faster
console.time('test')

const input = '16,1,0,18,12,14,19'.split(',').map(Number)
// const targetTurn = 2020 // Part 1
const targetTurn = 30000000 // Part 2

let currentValue = input[input.length - 1]
let currentTurn = input.length
let seen = new Uint32Array(targetTurn)

input.map((e,i) => seen[e] = i+1)

while (currentTurn < targetTurn) {
    if (seen[currentValue] !== 0) {
        let prevTurn = seen[currentValue]
        seen[currentValue] = currentTurn
        currentValue = currentTurn - prevTurn;
    } else {
        seen[currentValue] = currentTurn
        currentValue = 0
    }

    currentTurn++
}
console.log(currentValue)

console.timeEnd('test')

File: Day15/solution.py
input = [16, 1, 0, 18, 12, 14, 19]
# target_turn = 2020 #Part 1
target_turn = 30000000  # Part 2

current_value = input[-1]
current_turn = len(input)
seen = {}

for turn, num in enumerate(input):
    seen[num] = turn+1

while current_turn < target_turn:
    if current_value not in seen:
        seen[current_value] = current_turn
        current_value = 0
    else:
        last_turn = seen[current_value]
        seen[current_value] = current_turn
        current_value = current_turn - last_turn
    current_turn += 1

print(current_value)
